### PR TITLE
Upgrade to angular 16

### LINKED ISSUE #2752: Upgrade to Angular 16 or 17
Angular 15 LTS ends on 2024-05-18: https://angular.io/guide/releases#actively-supported-versions

We should use this opportunity to upgrade to the latest version of Angular (v17).  See https://update.angular.io/ for details

**Dependencies:**
As discussed in today's DevMtg, this work will be **dependent on** #2750 , described in more detail at https://wiki.lyrasis.org/display/DSPACE/Migration+to+Standalone+Components

Any work on this upgrade should be based off that PR 2750 because Angular 17 recommends standalone components.  

**Related Work:**
Several past Angular upgrades can be used as a reference, if necessary
* https://github.com/DSpace/dspace-angular/pull/2204
* https://github.com/DSpace/dspace-angular/pull/1567

### PR DESCRIPTION
**Pull request by atarix83**
## References
* Fixes #2752

## Description
This PR provide upgrade to angular 16

## Instructions for Reviewers
List of changes in this PR:
* since view engine libraries are no longer supported we had to : 
  * remove `ngx-sortablej` and replaced it by usig the [angular cdk drag/drop module](https://material.angular.io/cdk/drag-drop/overview).
  * remove `markdown-it-mathjax3`. Since this was only a plugin to support the MathJax for the markdown-it library we decide to use directly the MathJax library. It's not a direct dependency but we've [implemented a service](https://github.com/4Science/dspace-angular/blob/DURACOM-234/src/app/core/shared/client-math.service.ts) for injecting the script by retrieving it from a CDN. In this way we don't need to add to the bundle a so heavy dependency.   
* Node 16 is no longer supported so i've also changed the github workflow in order to run CI for version 18 and 20 of node

**Include guidance for how to test or review your PR.**
Review this PR by running the application (against sandbox rest server) to verify everything works properly as the current sandbox environment.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by atarix83** [comment]
@tdonohue @artlowel 

This PR is still WIP because there is a problem with the new version of NGRX. The problem is described well in this [discussion](https://github.com/ngrx/platform/discussions/3903) so currently our code lead to have several runtime errors like the following and i still need to understand if there is a way to avoid it. 

```
ERROR Error: NG0600: Writing to signals is not allowed in a `computed` or an `effect` by default. Use `allowSignalWrites` in the `CreateEffectOptions` to enable this inside effects.
    at core.mjs:28103:15
    at throwInvalidWriteToSignalError (core.mjs:2787:5)
    at Function.signalSetFn [as set] (core.mjs:2843:9)
    at Object.next (rxjs-interop.mjs:762:34)
    at ConsumerObserver.next (Subscriber.js:91:33)
    at SafeSubscriber._next (Subscriber.js:60:26)
    at SafeSubscriber.next (Subscriber.js:31:18)
    at map.js:7:24
    at OperatorSubscriber._next (OperatorSubscriber.js:13:21)
    at OperatorSubscriber.next (Subscriber.js:31:18)
```

Schermata del 2024-03-21 19-04-00

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks @atarix83!

I did a code review. The main issue I found was removing interfaces for resolvers and guards, because they're deprecated. I realize it's a lot of work to refactor all guards and resolvers as functions. However it looks like they're deprecated, but not removed yet. If you don't have the time to refactor them all now, I'd simply put them back the way they were so we don't lose that deprecation warning.

File: src/config/default-app-config.ts [hunk 1]
```diff
@@ -476,8 +476,8 @@ export class DefaultAppConfig implements AppConfig {
   // Whether to enable Markdown (https://commonmark.org/) and MathJax (https://www.mathjax.org/)
   // display in supported metadata fields. By default, only dc.description.abstract is supported.
   markdown: MarkdownConfig = {
-    enabled: false,
-    mathjax: false,
+    enabled: true,
+    mathjax: true,
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply src/config/default-app-config.ts:480]: I agree these should remain disabled.  I'm OK with either moving `mathjax` to its own setting or keeping it here for now.  I realize it's not directly related to markdown anymore, but the two concepts are "similar" in nature.

**Review by tdonohue — COMMENTED** [review]
Thanks @atarix83 !  I gave this a test today and I'm not finding any issues locally.  So, I'm basically a +1 once two things are cleaned up:

* First, you should obviously address @artlowel 's feedback
* Second, I notice the most recent bug fix has caused a single e2e test (in `submission.cy.ts`) to fail.  I'm not exactly sure why that is, so I restarted the tests to see if it's a random failure.  Ideally, we should get this to work again, obviously.  But, if it continues to fail with no known cause, we could disable the test temporarily if an immediate fix cannot be found & fix it in a follow-up PR.

File: src/config/default-app-config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (atarix83) [reply src/config/default-app-config.ts:480]: yeah, committed by mistake i'll fix it

**Comment by tdonohue** [comment]
@atarix83 : It looked like maybe you had this already updated, but I see it has some build errors (`yarn build:prod` fails).  Once it is building again & ready for review, please let @artlowel and I know so that we can review it quickly.

**Comment by tdonohue** [comment]
@atarix83 : It looks like e2e tests are still failing here.  Also, if you look at the logs of the e2e tests, there's some odd errors appearing like these:
```
ERROR Error [NullInjectorError]: R3InjectorError(Standalone[EditItemPageComponent2])[(...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params) -> (...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params) -> (...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params) -> (...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)]: 
  NullInjectorError: No provider for (...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)!
    at NullInjector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3574700)
    at R3Injector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3580383)
    at R3Injector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3580383)
    at R3Injector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3580383)
    at R3Injector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3580383)
    at ChainedInjector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3631994)
    at lookupTokenUsingModuleInjector (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3538532)
    at getOrCreateInjectable (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3540000)
    at NodeInjector.get (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3545455)
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/8725.js:1:44687 {
  ngTempTokenPath: null,
  ngTokenPath: [
    '(...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)',
    '(...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)',
    '(...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)',
    '(...params)=>(0,_angular_core__WEBPACK_IMPORTED_MODULE_1__.f3M)(provider).canActivate(...params)'
  ]
}
```
and
```
ERROR TypeError: Cannot read properties of undefined (reading 'length')
    at template (/home/runner/work/dspace-angular/dspace-angular/dist/server/8725.js:1:46618)
    at executeTemplate (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3605582)
    at refreshView (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3625885)
    at detectChangesInView (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3629054)
    at detectChangesInComponent (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3628753)
    at detectChangesInChildComponents (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3629404)
    at refreshView (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3627901)
    at detectChangesInView (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3629054)
    at detectChangesInEmbeddedViews (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3628645)
    at refreshView (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3626615)
```
and 
```
ERROR Error: Uncaught (in promise): TypeError: Cannot read properties of undefined (reading 'totalElements')
TypeError: Cannot read properties of undefined (reading 'totalElements')
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:312473
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2793944
    at OperatorSubscriber2._this._next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2755765)
    at OperatorSubscriber2.Subscriber2.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2709329)
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2830806
    at OperatorSubscriber2._this._next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2755765)
    at OperatorSubscriber2.Subscriber2.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2709329)
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2786375
    at OperatorSubscriber2._this._next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2755765)
    at OperatorSubscriber2.Subscriber2.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:2709329)
    at resolvePromise (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3052869)
    at resolvePromise (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3051848)
    at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3054734
    at _ZoneDelegate.invokeTask (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3038848)
    at Object.onInvokeTask (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3595930)
    at _ZoneDelegate.invokeTask (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3038738)
    at Zone3.runTask (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3031840)
    at drainMicroTaskQueue (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3042081)
    at invokeTask (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3040324)
    at ZoneTask.invoke (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3040105) {
```

It implies to me that we still have a few errors in SSR perhaps.  Any thoughts on whether you can try to fix these quickly?  You should be able to run these e2e tests via `yarn e2e` locally.

**Comment by atarix83** [comment]
sorry @tdonohue the PR was not yet ready to be reviewed again. Now it should if all tests pass.

@artlowel @tdonohue 
Regarding the deprecated interfaces for revolvers and guards, I've applied the following changes: 
- all revolvers have been migrated in order to use functional resolvers
- all the guards which didn't use class inheritance have been migrated in order to use functional guards
- for all the other guards whose migration would have taken too long due to the use of class inheritance (the ones which extend `SomeFeatureAuthorizationGuard` class) i've use the helper function [mapToCanActivate](https://angular.io/api/router/mapToCanActivate#description) provided by Angular. I suggest to open a follow-up task for migrating all of them to functional guards if you agree

**Comment by tdonohue** [comment]
NOTE: I *think* the e2e test failures are now because of CSRF issues with our Cypress test suite.   Now that https://github.com/DSpace/DSpace/pull/9321 is merged on the backend, I think that slightly changed the CSRF behavior (the changes were caused by Spring Security 6, not by us).  This difference in behavior appears to be causing Cypress to fail whenever it tries to login.

So, I think the e2e test failures here are NOT an issue with this PR as they are not reproducible in the UI.  Instead it is an issue with our Cypress test suite -- I'm looking into it, but I think it can be ignored for now.  If we have to we could temporarily disable the failing Cypress tests while we fix them.

**Review by artlowel — APPROVED** [review]
Thanks @atarix83!

>for all the other guards whose migration would have taken too long due to the use of class inheritance (the ones which extend `SomeFeatureAuthorizationGuard` class) i've use the helper function [mapToCanActivate](https://angular.io/api/router/mapToCanActivate#description) provided by Angular. I suggest to open a follow-up task for migrating all of them to functional guards if you agree

That's fine

One more thing though: the convention is for function names to start with a lowercase letter, and class names to start with a capital letter. All resolver-functions in this PR still start with a capital letter. If you can still fix it, that would be great, or you can add it to the ticket to clean up the guards.

**Comment by tdonohue** [comment]
@atarix83 and @artlowel : I've verified that the remaining e2e test failures in this PR **also exist** on `main`.  They are also visible on this new PR from dependabot: #2881

I'm working on a fix to these issues.  The problem is that the CSRF token is not being initialized properly **before** the e2e tests run, which causes the e2e tests to not be able to login (i.e. all tests that require authentication are failing).

The fix is to pre-initialize the CSRF token.. I'm borrowing code from #2838 (created by @artlowel 's team) and updating it to fix the issue.  With these fixes, the e2e tests are working now on my end, but I'm doing some minor code cleanup & I'll create a PR for this shortly.

**Comment by atarix83** [comment]
> One more thing though: the convention is for function names to start with a lowercase letter, and class names to start with a capital letter. All resolver-functions in this PR still start with a capital letter. If you can still fix it, that would be great, or you can add it to the ticket to clean up the guards.

@artlowel  done

**Comment by tdonohue** [comment]
@atarix83 : I've just merged #2886 which fixes the e2e test errors.  If you rebase this PR on the latest `main` (and fix the minor merge conflicts), your e2e tests should now succeed.

**Comment by atarix83** [comment]
@tdonohue  
I've aligned this on main branch.

@artlowel @tdonohue 
Looking at the #2886  i'm wondering if the code added [here](https://github.com/DSpace/dspace-angular/pull/2886/files#diff-c4b88760ec5c5d10957001fc56d12850dcf476a5802f52acdb5a6f14d37612bf) should be refactored in a better way.
I think the better place to do that check is the effect related to the dispatched ngrx action (https://github.com/DSpace/dspace-angular/blob/main/src/app/core/data/request.effects.ts#L43). In this way we avoid to add a nested subscription. 
@artlowel  do you agree? if so i can do the change in this PR

**Comment by artlowel** [comment]
@atarix83 I agree that refactor would be a good idea, However I wouldn't hold up this PR to do it. I'd rather do it in a small bugfix PR next week.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @atarix83 !  As discussed in today's dev meeting, I'm merging this immediately. The code looks good & I've verified it's working.  The only outstanding issues are the [refactor you mentioned above](https://github.com/DSpace/dspace-angular/pull/2871#issuecomment-2036469013) and the [potential security vulnerability reported by CodeQL](https://github.com/DSpace/dspace-angular/pull/2871/checks?check_run_id=23429832711).  I'm going to move those both to follow-up tickets & assign to you so that they can be fixed prior to 8.0 final.

**Pull request merged**